File: packages/oxlint-plugin-react-doctor/src/plugin/rules/a11y/no-noninteractive-element-to-interactive-role.ts
import { HTML_TAGS } from "../../constants/html-tags.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getElementType } from "../../utils/get-element-type.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";
import { isInteractiveRole } from "../../utils/is-interactive-role.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { isNonInteractiveElement } from "../../utils/is-non-interactive-element.js";
import type { Rule } from "../../utils/rule.js";

const buildMessage = (tag: string, role: string): string =>
  `Role \`${role}\` gives \`<${tag}>\` interactive semantics even though the element is noninteractive, so screen reader users get the wrong controls.`;

const DEFAULT_ALLOWED_ROLES: Record<string, ReadonlyArray<string>> = {
  ul: ["menu", "menubar", "radiogroup", "tablist", "tree", "treegrid"],
  ol: ["menu", "menubar", "radiogroup", "tablist", "tree", "treegrid"],
  li: ["menuitem", "menuitemcheckbox", "menuitemradio", "row", "tab", "treeitem"],
  fieldset: ["radiogroup", "presentation"],
};

interface NoNoninteractiveElementToInteractiveRoleSettings {
  // Element → list of allowed roles. Overrides the default table.
  [tagName: string]: ReadonlyArray<string> | undefined;
}

const resolveSettings = (
  settings: Readonly<Record<string, unknown>> | undefined,
): Record<string, ReadonlyArray<string>> => {
  const reactDoctor = settings?.["react-doctor"];
  if (typeof reactDoctor !== "object" || reactDoctor === null) return DEFAULT_ALLOWED_ROLES;
  const reactDoctorBlock = reactDoctor as {
    noNoninteractiveElementToInteractiveRole?: NoNoninteractiveElementToInteractiveRoleSettings;
  };
  if (!("noNoninteractiveElementToInteractiveRole" in reactDoctorBlock)) {
    return DEFAULT_ALLOWED_ROLES;
  }
  const ruleSettings = reactDoctorBlock.noNoninteractiveElementToInteractiveRole ?? {};
  // Explicit (possibly empty) override — replaces defaults verbatim.
  const result: Record<string, ReadonlyArray<string>> = {};
  for (const key of Object.keys(ruleSettings)) {
    const value = ruleSettings[key];
    if (Array.isArray(value)) result[key] = value;
  }
  return result;
};

// Port of `oxc_linter::rules::jsx_a11y::no_noninteractive_element_to_interactive_role`.
export const noNoninteractiveElementToInteractiveRole = defineRule<Rule>({
  id: "no-noninteractive-element-to-interactive-role",
  title: "Noninteractive element given interactive role",
  tags: ["react-jsx-only"],
  severity: "warn",
  recommendation:
    "Use a real interactive element instead of adding an interactive role to a static one.",
  category: "Accessibility",
  create: (context) => {
    const allowedRoles = resolveSettings(context.settings);
    return {
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        const roleAttribute = hasJsxPropIgnoreCase(node.attributes, "role");
        if (!roleAttribute) return;
        const value = roleAttribute.value as EsTreeNode | null;
        if (!value || !isNodeOfType(value, "Literal") || typeof value.value !== "string") {
          return;
        }
        const trimmed = value.value.trim();
        const firstRole = trimmed.split(/\s+/)[0];
        if (!firstRole) return;

        const elementType = getElementType(node, context.settings);
        if (!HTML_TAGS.has(elementType)) return;

        const allowed = allowedRoles[elementType];
        if (allowed && allowed.includes(firstRole)) return;
        if (!isNonInteractiveElement(elementType, node)) return;
        if (!isInteractiveRole(firstRole)) return;
        context.report({
          node: roleAttribute,
          message: buildMessage(elementType, firstRole),
        });
      },
    };
  },
});

File: native/oxlint/rules/no-noninteractive-element-to-interactive-role.rs
use oxc_ast::{AstKind, ast::JSXAttributeValue};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode,
    context::LintContext,
    globals::HTML_TAG,
    rule::Rule,
    utils::{
        get_element_type, has_jsx_prop_ignore_case, is_interactive_role, is_non_interactive_element,
    },
};

#[derive(Debug, Default, Clone)]
pub struct NoNoninteractiveElementToInteractiveRole;

declare_oxc_lint!(
    /// Prevent noninteractive elements from receiving interactive roles.
    NoNoninteractiveElementToInteractiveRole,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Prevent noninteractive elements from receiving interactive roles.",
);

impl Rule for NoNoninteractiveElementToInteractiveRole {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        let Some(oxc_ast::ast::JSXAttributeItem::Attribute(role_attribute)) =
            has_jsx_prop_ignore_case(opening_element, "role")
        else {
            return;
        };
        let Some(JSXAttributeValue::StringLiteral(role_value)) = role_attribute.value.as_ref()
        else {
            return;
        };
        let Some(role) = role_value.value.trim().split_whitespace().next() else {
            return;
        };
        let element_type = get_element_type(ctx, opening_element);
        if !HTML_TAG.contains(element_type.as_ref())
            || is_allowed_role(ctx, element_type.as_ref(), role)
            || !is_non_interactive_element(&element_type, opening_element)
            || !is_interactive_role(role)
            || (role == "separator"
                && has_jsx_prop_ignore_case(opening_element, "tabindex").is_none())
        {
            return;
        }
        let message = format!(
            "Role `{role}` gives `<{element_type}>` interactive semantics even though the element is noninteractive, so screen reader users get the wrong controls."
        );
        ctx.diagnostic(OxcDiagnostic::warn(message).with_label(role_attribute.span));
    }
}

fn is_allowed_role(ctx: &LintContext, element_type: &str, role: &str) -> bool {
    let rule_settings = ctx
        .settings()
        .json
        .as_ref()
        .and_then(|settings| settings.get("react-doctor"))
        .and_then(serde_json::Value::as_object)
        .and_then(|settings| settings.get("noNoninteractiveElementToInteractiveRole"));
    if let Some(rule_settings) = rule_settings {
        return rule_settings
            .as_object()
            .and_then(|settings| settings.get(element_type))
            .and_then(serde_json::Value::as_array)
            .is_some_and(|roles| roles.iter().any(|allowed| allowed.as_str() == Some(role)));
    }
    match element_type {
        "ul" | "ol" => matches!(
            role,
            "listbox" | "menu" | "menubar" | "radiogroup" | "tablist" | "tree" | "treegrid"
        ),
        "li" => matches!(
            role,
            "menuitem"
                | "menuitemcheckbox"
                | "menuitemradio"
                | "option"
                | "row"
                | "tab"
                | "treeitem"
        ),
        "table" => role == "grid",
        "td" => role == "gridcell",
        "nav" => role == "tablist",
        "fieldset" => matches!(role, "radiogroup" | "presentation"),
        _ => false,
    }
}
